### PR TITLE
CDAP-18559 : throwing RetryableProvisionException for read time out while fetching networks

### PR DESCRIPTION
**Pull request by sahusanket**

### CODE REVIEW

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocClient.java [hunk 1]
```diff
@@ -1076,6 +1099,32 @@ private RetryableProvisionException handleApiException(ApiException e) throws Re
     throw new DataprocRuntimeException(e);
   }
 
+  //Throws retryable Exception for the cases that are transient in nature
+  private static void handleComputeExceptions(Exception e) throws RetryableProvisionException {
+    // if there was an SocketTimeoutException ( read time out ) , we can just try again
+    if (e instanceof SocketTimeoutException) {
```
Comment (janakigg) [inline cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocClient.java:1112]: Does this cover 502 (gateway timeout)? 503 is also retryable https://softwareengineering.stackexchange.com/questions/341518/rest-how-to-determine-transient-exceptions

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocClient.java [hunk 2]
```diff
@@ -1076,6 +1099,32 @@ private RetryableProvisionException handleApiException(ApiException e) throws Re
     throw new DataprocRuntimeException(e);
   }
 
+  //Throws retryable Exception for the cases that are transient in nature
+  private static void handleComputeExceptions(Exception e) throws RetryableProvisionException {
```
Comment (janakigg) [inline cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocClient.java]: Nit: function name could be something else Ex. handleRetryableExceptions

**Comment by janakigg** [comment]
What unit tests can you add for this? Should we have integration tests for this featue?

File: cdap-runtime-ext-dataproc/pom.xml [hunk 3]
```diff
@@ -135,6 +135,14 @@
       <groupId>org.mockito</groupId>
       <artifactId>mockito-core</artifactId>
     </dependency>
+    <dependency>
+      <groupId>org.powermock</groupId>
```
Comment (tivv) [inline cdap-runtime-ext-dataproc/pom.xml:139]: Why do we need both Mockito and Powermock?

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocClient.java [hunk 4]
```diff
@@ -1076,6 +1099,37 @@ private RetryableProvisionException handleApiException(ApiException e) throws Re
...
+    if (e instanceof SocketTimeoutException) {
+      throw new RetryableProvisionException(e);
+    }
+
+    //Attempt retry in case of rate limit errors and service unavailable
+    if (e instanceof GoogleJsonResponseException) {
+      GoogleJsonResponseException gError = ((GoogleJsonResponseException) e);
+      int statusCode = gError.getStatusCode();
+
+      if (statusCode == 503) {
```
Comment (tivv) [inline cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocClient.java]: Please use proper constants for status codes, e.g. `java.net.HttpURLConnection#HTTP_UNAVAILABLE` here.

File: cdap-runtime-ext-dataproc/pom.xml (hunk 3 above)
Comment (tivv) [reply cdap-runtime-ext-dataproc/pom.xml:139]: Oh wow, we are using 8 year old Mockito version. Could you try bumping it to newer one, so we don't need powermock for static calls?

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocClient.java [hunk 5]
```diff
@@ -1076,6 +1101,37 @@ private RetryableProvisionException handleApiException(ApiException e) throws Re
...
+    //Attempt retry in case of rate limit errors and service unavailable
+    if (e instanceof GoogleJsonResponseException) {
+      GoogleJsonResponseException gError = ((GoogleJsonResponseException) e);
+      int statusCode = gError.getStatusCode();
+
+      if (statusCode == HttpURLConnection.HTTP_UNAVAILABLE) {
+        throw new RetryableProvisionException(e);
+      }
+
+      if (statusCode == HttpURLConnection.HTTP_FORBIDDEN || statusCode == 429) {
```
Comment (tivv) [inline cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocClient.java]: Please introduce a constant for 429 as well

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocClient.java [hunk 6]
```diff
@@ -1076,6 +1101,37 @@ private RetryableProvisionException handleApiException(ApiException e) throws Re
...
+        boolean isRetryAble = gError.getDetails().getErrors().stream()
+          .anyMatch(errorInfo ->
+                      errorInfo.getReason().equalsIgnoreCase("rateLimitExceeded") ||
```
Comment (tivv) [inline cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocClient.java]: Reason is nullable in ErrorInfo. Please reverse the check to handle null ("rateLimitExceeded".equalsIgnoreCase). BTW: Do you really need to ignore case? If not, you can just have a short set of valid codes and use contains

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocClient.java [hunk 7]
```diff
@@ -1088,4 +1141,9 @@ private RetryableProvisionException handleApiException(ApiException e) throws Re
       this.port = port;
     }
   }
+
+  private static Set<String> errorInfoReasons = Stream.of(
```
Comment (tivv) [inline cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocClient.java]: Use immutable set for any static members. Also use uppercase underscored names for constants and move to the beginning of the class

**Review by tivv — APPROVED** [review]

**Pull request merged**